File: lib/java/src/main/java/io/github/brenbar/uapi/internal/ConstructBinaryEncoding.java
package io.github.brenbar.uapi.internal;

import java.util.HashMap;
import java.util.Map;
import java.util.TreeSet;

import io.github.brenbar.uapi.UApiSchema;

import static io.github.brenbar.uapi.internal.CreateChecksum.createChecksum;

public class ConstructBinaryEncoding {
    public static _BinaryEncoding constructBinaryEncoding(UApiSchema uApiSchema) {
        final var allKeys = new TreeSet<String>();
        for (final var entry : uApiSchema.parsed.entrySet()) {
            allKeys.add(entry.getKey());

            if (entry.getValue() instanceof final _UStruct s) {
                final Map<String, _UFieldDeclaration> structFields = s.fields;
                allKeys.addAll(structFields.keySet());
            } else if (entry.getValue() instanceof final _UUnion u) {
                final Map<String, _UStruct> unionCases = u.cases;
                for (final var entry2 : unionCases.entrySet()) {
                    allKeys.add(entry2.getKey());
                    final var struct = entry2.getValue();
                    final var structFields = struct.fields;
                    allKeys.addAll(structFields.keySet());
                }
            } else if (entry.getValue() instanceof final _UFn f) {
                final _UUnion fnCall = f.call;
                final Map<String, _UStruct> fnCallCases = fnCall.cases;
                final _UUnion fnResult = f.result;
                final Map<String, _UStruct> fnResultCases = fnResult.cases;

                for (final var e2 : fnCallCases.entrySet()) {
                    allKeys.add(e2.getKey());
                    final var struct = e2.getValue();
                    final Map<String, _UFieldDeclaration> structFields = struct.fields;
                    allKeys.addAll(structFields.keySet());
                }

                for (var e2 : fnResultCases.entrySet()) {
                    allKeys.add(e2.getKey());
                    var struct = e2.getValue();
                    final Map<String, _UFieldDeclaration> structFields = struct.fields;
                    allKeys.addAll(structFields.keySet());
                }
            }
        }
        var i = 0;
        final var binaryEncodingMap = new HashMap<String, Integer>();
        for (final var key : allKeys) {
            binaryEncodingMap.put(key, i);
            i += 1;
        }
        final var finalString = String.join("\n", allKeys);

        final int checksum = createChecksum(finalString);
        return new _BinaryEncoding(binaryEncodingMap, checksum);
    }
}

File: lib/ts/uapi/internal/binary/ConstructBinaryEncoding.ts
import { UApiSchema } from 'uapi/UApiSchema';
import { BinaryEncoding } from 'uapi/internal/binary/BinaryEncoding';
import { createChecksum } from 'uapi/internal/binary/CreateChecksum';
import { UUnion } from 'uapi/internal/types/UUnion';
import { UStruct } from 'uapi/internal/types/UStruct';
import { UFn } from 'uapi/internal/types/UFn';
import { UFieldDeclaration } from 'uapi/internal/types/UFieldDeclaration';

export function constructBinaryEncoding(uApiSchema: UApiSchema): BinaryEncoding {
    const allKeys: Set<string> = new Set();

    for (const [key, value] of Object.entries(uApiSchema.parsed)) {
        allKeys.add(key);

        if (value instanceof UStruct) {
            const structFields: Record<string, UFieldDeclaration> = value.fields;
            for (const structFieldKey of Object.keys(structFields)) {
                allKeys.add(structFieldKey);
            }
        } else if (value instanceof UUnion) {
            const unionCases: Record<string, UStruct> = value.cases;
            for (const [caseKey, caseValue] of Object.entries(unionCases)) {
                allKeys.add(caseKey);
                const structFields = caseValue.fields;
                for (const structFieldKey of Object.keys(structFields)) {
                    allKeys.add(structFieldKey);
                }
            }
        } else if (value instanceof UFn) {
            const fnCallCases: Record<string, UStruct> = value.call.cases;
            const fnResultCases: Record<string, UStruct> = value.result.cases;

            for (const [caseKey, caseValue] of Object.entries(fnCallCases)) {
                allKeys.add(caseKey);
                const structFields = caseValue.fields;
                for (const structFieldKey of Object.keys(structFields)) {
                    allKeys.add(structFieldKey);
                }
            }

            for (const [caseKey, caseValue] of Object.entries(fnResultCases)) {
                allKeys.add(caseKey);
                const structFields = caseValue.fields;
                for (const structFieldKey of Object.keys(structFields)) {
                    allKeys.add(structFieldKey);
                }
            }
        }
    }

    const sortedAllKeys = Array.from(allKeys).sort();
    const binaryEncoding = new Map<string, number>();
    sortedAllKeys.forEach((key, index) => {
        binaryEncoding.set(key, index);
    });

    const finalString = sortedAllKeys.join('\n');
    const checksum = createChecksum(finalString);

    return new BinaryEncoding(binaryEncoding, checksum);
}
